### PR TITLE
Add logic to recipient submission widget

### PR DESCRIPTION
**Pull request by corwintines**
https://github.com/ethereum/clrfund/issues/67

### CODE REVIEW

File: vue-app/src/App.vue [hunk 1]
```diff
@@ -80,8 +80,8 @@ export default class App extends Vue {
     setInterval(() => {
       this.$store.dispatch(LOAD_ROUND_INFO)
     }, 60 * 1000)
-    setInterval(() => {
-      this.$store.dispatch(LOAD_RECIPIENT_REGISTRY_INFO)
+    setInterval(async () => {
+      await this.$store.dispatch(LOAD_RECIPIENT_REGISTRY_INFO)
```
Comment (samajammin) [inline vue-app/src/App.vue:84]: Why should a dispatch be async?

File: vue-app/src/components/RecipientSubmissionWidget.vue [hunk 2]
```diff
@@ -165,67 +169,120 @@ export default class RecipientSubmissionWidget extends Vue {
...
+        this.isWaiting = false
+        this.txError = error.message
+        return
+      }
       this.isWaiting = false
-      this.txError = error.message
-      return
+
+      this.$router.push({
+        name: 'projectAdded',
```
Comment (pettinarip) [inline vue-app/src/components/RecipientSubmissionWidget.vue]: I've changed the router names. This should be `'project-added'`

File: vue-app/src/store/index.ts [hunk 3]
```diff
@@ -527,6 +524,19 @@ const getters = {
...
     )
   },
+  hasLowFunds(state: RootState): boolean {
+    const { currentUser, recipientRegistryInfo } = state
+
+    if (
+      currentUser &&
+      currentUser.balance &&
+      recipientRegistryInfo &&
+      recipientRegistryInfo.deposit
```
Comment (pettinarip) [inline vue-app/src/store/index.ts]: You could use Optional Chaining for this.
```ts
if (currentUser?.balance && recipientRegistryInfo?.deposit) {
  ...
}
```
ref https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining

File: vue-app/src/App.vue (hunk 1 above)
Comment (corwintines) [reply vue-app/src/App.vue:84]: I just saw that LOAD_RECIPIENT_REGISTRY_INFO was async, so was using the await for that.

File: vue-app/src/store/index.ts (hunk 3 above)
Comment (corwintines) [reply vue-app/src/store/index.ts]: Thanks for suggestion, missed this in cleanup.

File: vue-app/src/components/RecipientSubmissionWidget.vue (hunk 2 above)
Comment (corwintines) [reply vue-app/src/components/RecipientSubmissionWidget.vue]: Good catch, my mistake on merging in prettier stuff.

File: vue-app/src/App.vue (hunk 1 above)
Comment (samajammin) [reply vue-app/src/App.vue:84]: Ah, right, originally I was thinking no dispatches should be async but I was wrong about that... https://vuex.vuejs.org/guide/actions.html#composing-actions

File: vue-app/src/components/RecipientSubmissionWidget.vue [hunk 4]
```diff
@@ -165,67 +169,120 @@ export default class RecipientSubmissionWidget extends Vue {
...
+      recipient
+    ) {
+      const registry = new Contract(
+        recipientRegistryAddress,
+        OptimisticRecipientRegistry,
+        this.walletProvider
+      )
+
+      return await registry.estimateGas.addRecipient(
+        recipient.fund.address,
```
Comment (samajammin) [inline vue-app/src/components/RecipientSubmissionWidget.vue]: I believe this should be `recipientRegistryAddress`

Comment (samajammin) [reply vue-app/src/components/RecipientSubmissionWidget.vue]: In general, shouldn't the arguments be the same here?
```
          addRecipient(
            recipientRegistryAddress,
            recipient,
            recipientRegistryInfo.deposit,
            currentUser.walletProvider.getSigner()
          ),
```

File: vue-app/src/components/RecipientSubmissionWidget.vue [hunk 5]
```diff
@@ -165,67 +169,120 @@ export default class RecipientSubmissionWidget extends Vue {
...
+        JSON.stringify({
```
Comment (samajammin) [inline vue-app/src/components/RecipientSubmissionWidget.vue]: I think we want to use `formToRecipientData` from /api/recipient-registry-optimistic.ts here. See how we construct it here: https://github.com/ethereum/clrfund/blob/develop/vue-app/src/api/recipient-registry-optimistic.ts#L279

**Comment by samajammin** [comment]
Hmmmm haven't been able to inspect this close enough. The code looks mostly right but testing it out, the gas estimate is far off:

Image 2021-06-25 at 7 24 26 AM

**Comment by corwintines** [comment]
> Hmmmm haven't been able to inspect this close enough. The code looks mostly right but testing it out, the gas estimate is far off:
> 
> Image 2021-06-25 at 7 24 26 AM

@samajammin I think I have an idea whats going on here, need to getGasPrice still and add that multiplication in. Missed doing this for some reason. My bad, will get that pushed up today.

**Review by corwintines — COMMENTED** [review]
Note: commented out the logic for trying to deal with the gas fee estimation. Took that issue out of the description for this PR, so it should only be addressing https://github.com/ethereum/clrfund/issues/67

File: vue-app/src/store/index.ts [hunk 6]
```diff
@@ -527,6 +524,14 @@ const getters = {
       (getters.isRoundContributionPhase || getters.isRoundReallocationPhase)
     )
   },
+  hasLowFunds(state: RootState): boolean {
```
Comment (samajammin) [inline vue-app/src/store/index.ts]: This should compare `currentUser.etherBalance` if `recipientRegistryInfo.depositToken` is ETH.

Comment (samajammin) [reply vue-app/src/store/index.ts]: I'll update.

File: vue-app/src/components/RecipientSubmissionWidget.vue [hunk 7]
```diff
@@ -133,26 +141,22 @@ import { waitForTransaction } from '@/utils/contracts'
 export default class RecipientSubmissionWidget extends Vue {
   @Prop() cta!: string
   @Prop() pending!: string
-  isLoading = false
+  isLoading = true
   isWaiting = false // TODO add logic
```
Comment (samajammin) [inline vue-app/src/components/RecipientSubmissionWidget.vue:142]: Did you check if this `isWaiting` logic is correct?

Comment (samajammin) [reply vue-app/src/components/RecipientSubmissionWidget.vue:142]: Ah I see we've moved this to #168

File: vue-app/src/store/index.ts (hunk 6 above)
Comment (samajammin) [reply vue-app/src/store/index.ts]: @corwintines I updated this but am realizing we may just want this as a getter in the RecipientSubmissionWidget component as it depends on `recipientRegistryInfo.depositToken`.

Other checks for `hasLowFunds` will depend on other criteria, e.g. on when contributing we should check for low DAI & ETH balance.

Comment (corwintines) [reply vue-app/src/store/index.ts]: Thats fair, I had that originally and then tried to abstract it out incase we were going to use it in other places. But I agree, there is probably a better case to have this done in the RecipientSubmissionWidget. I'll refactor that back in.

Comment (corwintines) [reply vue-app/src/store/index.ts]: Ive removed this getter and moved the low funds check into the RecipientSubmissionWidget. Can add the getter at a later time if there is a use for it.

File: vue-app/src/components/RecipientSubmissionWidget.vue [hunk 8]
```diff
@@ -72,20 +69,30 @@
...
+              }}{{ calculateFiatFee(this.estimatedGasFee) }})</span
+            >
+          </p>
         </div> -->
         <div class="cta">
           <button
             @click="handleSubmit"
             :class="
-              isWaiting || isPending || hasLowFunds
+              isWaiting || isPending || $store.getters.hasLowFunds
```
Comment (samajammin) [inline vue-app/src/components/RecipientSubmissionWidget.vue]:
```suggestion
              isWaiting || isPending || hasLowFunds
```

File: vue-app/src/components/RecipientSubmissionWidget.vue [hunk 9]
```diff
@@ -165,67 +166,129 @@ export default class RecipientSubmissionWidget extends Vue {
...
-      ? formatAmount(this.registryInfo.deposit, 18)
+    return this.$store.state.recipientRegistryInfo
+      ? formatAmount(this.$store.state.recipientRegistryInfo.deposit, 18)
       : '...'
   }
 
+  get hasLowFunds(): boolean {
+    const { currentUser, recipientRegistryInfo } = this.$store.state
+
+    if (currentUser?.balance && recipientRegistryInfo?.deposit) {
```
Comment (samajammin) [inline vue-app/src/components/RecipientSubmissionWidget.vue]:
```suggestion
    if (currentUser?.etherBalance && recipientRegistryInfo?.deposit) {
```

File: vue-app/src/components/RecipientSubmissionWidget.vue [hunk 10]
```diff
@@ -165,67 +166,129 @@ export default class RecipientSubmissionWidget extends Vue {
...
+      return currentUser.balance.lt(recipientRegistryInfo.deposit)
```
Comment (samajammin) [inline vue-app/src/components/RecipientSubmissionWidget.vue]:
```suggestion
      return currentUser.etherBalance.lt(recipientRegistryInfo.deposit)
```

File: vue-app/src/components/RecipientSubmissionWidget.vue [hunk 11]
```diff
@@ -133,26 +138,22 @@ import { waitForTransaction } from '@/utils/contracts'
...
-  // feeToken = 'ETH' // TODO add logic
-  // fiatFee = '1.04' // TODO add logic
-  // fiatSign = '$' // TODO add logic
-
-  registryInfo: RegistryInfo | null = null
+  ethPrice: EthPrice | null = null
+  fiatFee = '-'
+  fiatSign = '$'
+  // estimatedGasFee: BigNumber = BigNumber.from(0) TODO: Once either EIP-1559 or Arbitrum stuff is sorted, come back and finish gas estimate
+  feeToken = 'ETH'
```
Comment (samajammin) [inline vue-app/src/components/RecipientSubmissionWidget.vue]: Should we just use `depositToken` here?

**Review by samajammin — CHANGES_REQUESTED** [review]
A few final comments - feel free to merge after addressing these changes 👍

**Comment by samajammin** [comment]
Made those changes I mentioned 👍

**Pull request merged**